### ISSUE TITLE
Example using a trained model/policy

### ISSUE THREAD

**Issue by aliyakamercan** [deepmind/open_spiel#234]
Hi,

I am trying to run Kuhn poker playthrough using different policies (NSFP and CFR from examples), but I could not find a way to save these. Pickling doesn't seem to work. Is there some other method to save/load these?

Thanks

**Comment by lanctot** [comment]
Hi, can I double-check that you are doing this all in python? I think I know why the pickling does not work, and there is an easy fix (convert the policies to tabular policies and pickle those). We don't have an example but it would be easy to add. I will wait for your reply first.

**Comment by aliyakamercan** [comment]
I am using python.

Here are two examples:

```python
# 1
random_policy = policy.UniformRandomPolicy(game)
print(pickle.dumps(random_policy))

# 2
cfr_solver = cfr.CFRSolver(game)
iterations = 1000

for i in range(iterations):
  cfr_value = cfr_solver.evaluate_and_update_policy()
  print("Game util at iteration {}: {}".format(i, cfr_value))

cfr_policy = cfr_solver.average_policy()
print(pickle.dumps(random_policy))
```

Both pickles are failing with
```
can't pickle pyspiel.GameType objects
```

I think, it is because there is no `py::pickle` for `GameState` in [pyspiel.cc](https://github.com/deepmind/open_spiel/blob/master/open_spiel/python/pybind11/pyspiel.cc)

**Comment by lanctot** [comment]
That is odd because we definitely support pickling states, we even test it for all games in the simulation tests:

 https://github.com/deepmind/open_spiel/blob/0e9b1b75dea3e87f0326fe1012d1b0367196b7dc/open_spiel/python/tests/games_sim_test.py#L98

So something seems very wrong indeed. I will look into it. Did you perhaps add something to GameType that does not serialize?

**Comment by lanctot** [comment]
And the `py::pickle` is here:

**Comment by inejc** [comment]
I might be missing something here but I don't think there is support for serialization of any of the policy types currently, i.e. it is not possible to pickle `TabularPolicy` or `CFRAveragePolicy` which I believe the above python code tries to do.

**Comment by lanctot** [comment]
Oh I think you meant `GameType` which is missing the `py::pickle`. 

Yes it is probably right that we can't pickle tabular policies right now but these are simple objects. We should support pickling them and I think the fix is easy (or should be).

**Comment by inejc** [comment]
Once https://github.com/deepmind/open_spiel/pull/222 is finished, it should add the ability to pickle both `TabularPolicy` instances and the actual CFR solvers which can then be used to obtain the `CFRAveragePolicy`.

**Comment by lanctot** [comment]
Thanks @inejc but I thought you were doing the C++ side. I think/thought the python tabular policy objects were self-contained (i.e. not wrappers to the C++ TabularPolicy objects)... but maybe I am wrong?

**Comment by inejc** [comment]
> Thanks @inejc but I thought you were doing the C++ side.

That's correct. I didn't realize the code above uses a pure python version of tabular, sorry for the confusion!

**Comment by aliyakamercan** [comment]
Thanks for the quick replies.

> Oh I think you meant `GameType` which is missing the `py::pickle`.
> 
> Yes it is probably right that we can't pickle tabular policies right now but these are simple objects. We should support pickling them and I think the fix is easy (or should be).

Yep, my bad, I meant `GameType`. 

https://docs.python.org/3/library/pickle.html#handling-stateful-objects can be used to drop/set game_type since `game` is already serializable.

```diff
diff --git a/open_spiel/python/policy.py b/open_spiel/python/policy.py
index 75252fc..1bcf647 100644
--- a/open_spiel/python/policy.py
+++ b/open_spiel/python/policy.py
@@ -312,6 +312,14 @@ class TabularPolicy(Policy):
         1 - alpha) * probability_array + alpha * noise_mask
     return copied_instance
 
+  def __getstate__(self):
+    state = self.__dict__.copy()
+    del state['game_type']
+    return state
+
+  def __setstate__(self, state):
+    self.__dict__.update(state)
+    self.game_type = self.game.get_type()
 
 class UniformRandomPolicy(Policy):
   """Policy where the action distribution is uniform over all legal actions.
``` 

I will open a PR.

How about solvers like deep-CFR where I wouldn't necessarily want to store the tabular policy? Should I be using `tf.saver` in those cases and save the network?

**Comment by lanctot** [comment]
Ok great thanks. I still want GameType to be serializable so I might add that as well this week.

For other things like NFSP we only need tf.saver if you want a full checkpoint. If you are happy with just the policy itself than we can extract it into a tabular policy by doing a tree pass and calling the network at each infostate (we might already be doing this in some places but if not we should add a general helper function that does this).

**Comment by ayakamercan-coatue** [comment]
I will work on making `GameType` serializable, let's close #235.

**Comment by lanctot** [comment]
Hi @ayakamercan-coatue , I will assume you are the same person as @aliyakamercan-- otherwise I would not want to decide on closing another person's contribution! ;-) 

I've already got the GameType serialization half-done (it's actually non-trivial because of the serializing of the GameType parameters). I can commit what I have and let you finish it. Are you happy with that? That way you still get credit for the contribution and I can do other stuff. 

WDYT?

**Comment by lanctot** [comment]
> I will work on making `GameType` serializable, let's close #235.

Feel free to close it yourself, I agree -- the serializing of the GameType will get this already, and I think I'd rather that as well.

**Comment by aliyakamercan** [comment]
> Hi @ayakamercan-coatue , I will assume you are the same person as @aliyakamercan-- otherwise I would not want to decide on closing another person's contribution! ;-)
> 
> I've already started on the GameType serialization (it's actually non-trivial because of the serializing of the GameType parameters). I can commit what I have and let you finish it. Are you happy with that? That way you still get credit for the contribution and I can do other stuff.
> 
> WDYT?

Got my users mixed up :), yes that sounds great.

**Comment by lanctot** [comment]
Ok, great! It will be in an upcoming commit in the next hour or two.

Can I ask the following:

1. The game type be readable: one piece of it per line, that way it is easily splittable with '\n' as a delimiter using absl::StrSplit. For the parameter specs line it would be like: "parameter_specification: <serialized parm1> <serialized parm2> <serialized parm3> ..." (you can do all this in spiel.cc; add whatever helper methods you need local to that file)

2. Add the py::pickle lines necessary to pyspiel.cc

3. In the python/game_sims_test can you also add a line to verify that the game type serialized and deserializes, and that the deserialized game type is equal to the game.game_type() ? I think you can do this right next to where the game gets serialized and deserialized.

4. Add an example or test in python/examples that does what you showed in this thread originally? (If it's a test then prefer to read/write string data rather than saving files)

Appreciate this, thanks!
